--- bundle/nginx-1.21.4/src/core/nginx.h	2023-07-27 11:52:31.049992503 +0800
+++ build/nginx-1.21.4/src/core/nginx.h	2023-07-27 11:06:05.442714388 +0800
@@ -9,8 +9,8 @@
 #define _NGINX_H_INCLUDED_
 
 
-#define nginx_version      1021004
-#define NGINX_VERSION      "1.21.4"
+#define nginx_version      1025001
+#define NGINX_VERSION      "1.25.1"
 #define NGINX_VER          "openresty/" NGINX_VERSION ".1"
 
 #ifdef NGX_BUILD
